integrations/ollama/src/haystack_integrations/components/generators/ollama/chat/chat_generator.py

```diff
@@ -234,118 +270,168 @@ def from_dict(cls, data: Dict[str, Any]) -> "OllamaChatGenerator":
 
-        # Create the ChatMessage with the combined content
-        combined_text = "".join([c.content for c in chunks])
-        reply = ChatMessage.from_assistant(combined_text)
+        # For every tool-call id store EITHER
+        #   * a concatenated str (when the model streams JSON snippets), OR
+        #   * the most-recent dict (when it sends a full JSON object at once)
+        tool_call_args_by_id: Dict[str, Union[str, dict]] = {}
+        last_seen_calls: List[Dict[str, Any]] = []
+
+        for raw_chunk in response_iter:
+            chunk = self._build_chunk(raw_chunk)
+            chunks.append(chunk)
+
+            if streaming_callback is not None:
+                streaming_callback(chunk)
+
+            # ── collect tool-call deltas ──────────────────────────────
+            for tc in chunk.meta.get("tool_calls", []):
+                tc_id = tc.get("id") or tc["function"]["name"]
+                arg_delta = tc["function"].get("arguments")
+
+                if arg_delta is None:
+                    continue
+
+                prev = tool_call_args_by_id.get(tc_id)
+
+                if isinstance(arg_delta, str):
+                    if isinstance(prev, str):
+                        tool_call_args_by_id[tc_id] = prev + arg_delta
+                    else:  # prev is dict or None → start fresh with str
+                        tool_call_args_by_id[tc_id] = arg_delta
+                else:  # arg_delta is dict
+                    tool_call_args_by_id[tc_id] = arg_delta  # overwrite; latest wins
+
+                last_seen_calls.append(tc)
+
+        # ── merge text ────────────────────────────────────────────────
+        combined_text = "".join(c.content for c in chunks)
+
+        # ── rebuild completed ToolCall objects ────────────────────────
+        completed_tool_calls: List[ToolCall] = []
+        if last_seen_calls:
+            seen_ids = set()
+            # walk backwards so earlier duplicates are skipped
+            for tc in reversed(last_seen_calls):
+                tc_id = tc.get("id") or tc["function"]["name"]
+                if tc_id in seen_ids:
+                    continue
+                seen_ids.add(tc_id)
+
+                completed_tool_calls.append(
+                    ToolCall(
+                        tool_name=tc["function"]["name"],
+                        arguments=tool_call_args_by_id.get(tc_id),
+                    )
+                )
+            completed_tool_calls.reverse()  # restore original order
+
+        reply = ChatMessage.from_assistant(
+            text=combined_text,
+            tool_calls=completed_tool_calls or None,
+        )
 
         # Convert the last chunk's metadata to OpenAI format and attach it to the ChatMessage
         if chunks:
-            meta = _convert_ollama_meta_to_openai_format(chunks[-1].meta)
-            reply._meta = meta
+            reply._meta = _convert_ollama_meta_to_openai_format(chunks[-1].meta)
```

Comment:
I realized it was this way before but could we add this directly to the creation of `ChatMessage` so 
```python
        reply = ChatMessage.from_assistant(
            text=combined_text,
            tool_calls=completed_tool_calls or None,
            meta=_convert_ollama_meta_to_openai_format(chunks[-1].meta) if chunks else None
        )
```

Reply:
done